File: linkhive-backend/app/Models/Goal.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Goal extends Model
{
    use HasFactory;

    protected $fillable = [
        'user_id',
        'goal_type',
        'description',
        'target_count',
        'progress',
        'start_date',
        'end_date',
        'completed',
    ];

    protected $casts = [
        'start_date' => 'date',
        'end_date' => 'date',
        'completed' => 'boolean',
    ];

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function updateProgress(): void
    {
        // Logic to calculate progress based on link statuses
        $this->progress = min($this->target_count, $this->progress + 1);
        $this->completed = $this->progress >= $this->target_count;
        $this->save();
    }
}

File: linkhive_app/lib/models/goal.dart
class Goal {
  final int? id;
  final int userId;
  final String goalType;
  final String description;
  final int targetCount;
  final int progress;
  final DateTime startDate;
  final DateTime endDate;
  final bool completed;

  Goal({
    this.id,
    required this.userId,
    required this.goalType,
    required this.description,
    required this.targetCount,
    this.progress = 0,
    required this.startDate,
    required this.endDate,
    this.completed = false,
  });

  factory Goal.fromJson(Map<String, dynamic> json) {
    return Goal(
      id: json['id'],
      userId: json['user_id'],
      goalType: json['goal_type'],
      description: json['description'],
      targetCount: json['target_count'],
      progress: json['progress'] ?? 0,
      startDate: DateTime.parse(json['start_date']),
      endDate: DateTime.parse(json['end_date']),
      completed: json['completed'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'goal_type': goalType,
      'description': description,
      'target_count': targetCount,
      'progress': progress,
      'start_date': startDate.toIso8601String(),
      'end_date': endDate.toIso8601String(),
      'completed': completed,
    };
  }

  double get progressPercentage {
    if (targetCount == 0) return 0;
    return (progress / targetCount * 100).clamp(0, 100);
  }
}